src/main/haskell/language-kore/src/Data/Kore/ASTUtils/SmartConstructors.hs

```diff
@@ -0,0 +1,328 @@
+    { getSortVariable = noLocationId "__FlexibleSort__" } --FIXME
+
+pattern And_          s2   a b = Fix (AndPattern (And s2 a b))
+pattern App_ h c               = Fix (ApplicationPattern (Application h c))
+pattern Bottom_       s2       = Fix (BottomPattern (Bottom s2))
+pattern Ceil_      s1 s2   a   = Fix (CeilPattern (Ceil s1 s2 a))
+pattern DV_           s2   a   = Fix (DomainValuePattern (DomainValue s2 a))
+pattern Equals_    s1 s2   a b = Fix (EqualsPattern (Equals s1 s2 a b))
+pattern Exists_       s2 v a   = Fix (ExistsPattern (Exists s2 v a))
+pattern Floor_     s1 s2   a   = Fix (FloorPattern (Floor s1 s2 a))
+pattern Forall_       s2 v a   = Fix (ForallPattern (Forall s2 v a))
+pattern Iff_          s2   a b = Fix (IffPattern (Iff s2 a b))
+pattern Implies_      s2   a b = Fix (ImpliesPattern (Implies s2 a b))
+pattern In_        s1 s2   a b = Fix (InPattern (In s1 s2 a b))
+pattern Next_         s2   a   = Fix (NextPattern (Next s2 a)) 
+pattern Not_          s2   a   = Fix (NotPattern (Not s2 a))
+pattern Or_           s2   a b = Fix (OrPattern (Or s2 a b))
+pattern Rewrites_     s2   a b = Fix (RewritesPattern (Rewrites s2 a b))
+pattern Top_          s2       = Fix (TopPattern (Top s2))
+pattern Var_             v     = Fix (VariablePattern v)   
+
+patternLens
+  :: (Applicative f, MetaOrObject level) 
+  => (Sort level -> f (Sort level))
+  -> (Sort level -> f (Sort level))
+  -> (Variable level -> f (Variable level))
+  -> (CommonPurePattern level -> f (CommonPurePattern level))
+  -> (CommonPurePattern level -> f (CommonPurePattern level))
+patternLens 
+  i  --input sort
+  o -- result sort
+  var -- variable
+  c -- child
+  = \case 
+  And_       s2   a b -> And_      <$>          o s2           <*> c a <*> c b
+  Bottom_    s2       -> Bottom_   <$>          o s2
+  Ceil_   s1 s2   a   -> Ceil_     <$> i s1 <*> o s2           <*> c a
+  DV_        s2   a   -> DV_       <$>          o s2           <*> c a
+  Equals_ s1 s2   a b -> Equals_   <$> i s1 <*> o s2           <*> c a <*> c b
+  Exists_    s2 v a   -> Exists_   <$>          o s2 <*> var v <*> c a
+  Floor_  s1 s2   a   -> Floor_    <$> i s1 <*> o s2           <*> c a
+  Forall_    s2 v a   -> Forall_   <$>          o s2 <*> var v <*> c a
+  Iff_       s2   a b -> Iff_      <$>          o s2           <*> c a <*> c b
+  Implies_   s2   a b -> Implies_  <$>          o s2           <*> c a <*> c b
+  In_     s1 s2   a b -> In_       <$> i s1 <*> o s2           <*> c a <*> c b
+  Next_      s2   a   -> Next_     <$>          o s2           <*> c a
+  Not_       s2   a   -> Not_      <$>          o s2           <*> c a
+  Or_        s2   a b -> Or_       <$>          o s2           <*> c a <*> c b
+  Rewrites_  s2   a b -> Rewrites_ <$>          o s2           <*> c a <*> c b
+  Top_       s2       -> Top_      <$>          o s2
+  Var_          v     -> Var_      <$>                   var v
+  other -> error $ "Unknown constructor " ++ show other
```

Comment:
Pattern Synonyms + GADTS = GHC definitely can't tell if it's complete or not, so without the "other" it doesn't compile.

Reply:
Oh, wow, I was waiting to use these synonyms in my condition-evaluation pull request, but now I'm not so sure anymore.

One of the reasons we chose Haskell is because all of the built-in safety, but it seems to me that Haskell programmers don't usually want it that much.

Reply:
`PatternSynonyms` are a big problem for completeness checking. For our case, we really only want to be using the synonyms for construction, not matching. The benefit of using `Fix` is that we should be able to write algebras that only need to match on the underlying functior!

Reply:
Actually, in this case the COMPLETE pragma works. So writing a complete list of cases with 

Fix ( ApplicationPattern (Application ...
Fix (AndPattern (And ...

is equivalent to

App_ ...
And_ ...

and the compiler will know it's complete.